tsdb/tsm1/engine.go

```diff
@@ -1104,21 +1116,82 @@ func (e *Engine) compactFull(ctx context.Context, grp CompactionGroup, wg *sync.
 
 	// Try the lo priority limiter, otherwise steal a little from the high priority if we can.
 	if e.compactionLimiter.TryTake() {
+		// Attempt to get ownership of the semaphore for this engine. If the
+		// default semaphore is in use then ownership will always be granted.
+		ttl := influxdb.DefaultLeaseTTL
+		lastCompaction := time.Duration(atomic.LoadInt64(&e.lastFullCompactionDuration))
+		if lastCompaction > ttl {
+			ttl = lastCompaction // If the last full compaction took > default ttl then set a new TTL
+		}
+
+		lease, err := e.fullCompactionSemaphore.TryAcquire(ctx, ttl)
+		if err == influxdb.ErrNoAcquire {
+			e.logger.Info("Cannot acquire semaphore ownership to carry out full compaction", zap.Duration("semaphore_requested_ttl", ttl))
+			e.compactionLimiter.Release()
+			return false
+		} else if err != nil {
+			e.logger.Warn("Failed to execute full compaction", zap.Error(err), zap.Duration("semaphore_requested_ttl", ttl))
+			e.compactionLimiter.Release()
+			return false
+		} else if e.fullCompactionSemaphore != influxdb.NopSemaphore {
+			e.logger.Info("Acquired semaphore ownership for full compaction", zap.Duration("semaphore_requested_ttl", ttl))
+		}
+
+		done := e.monitorLease(ctx, lease) // Closed when compaction finished.
+
 		e.compactionTracker.IncFullActive()
 		wg.Add(1)
 		go func() {
 			defer wg.Done()
 			defer e.compactionTracker.DecFullActive()
 			defer e.compactionLimiter.Release()
+
+			now := time.Now() // Track how long compaction takes
 			s.Apply(ctx)
+			atomic.StoreInt64(&e.lastFullCompactionDuration, int64(time.Since(now)))
+
 			// Release the files in the compaction plan
 			e.CompactionPlan.Release([]CompactionGroup{s.group})
+			close(done)
 		}()
 		return true
 	}
 	return false
 }
 
+// monitorLease initialises a goroutine that keeps a lease alive until the
+// returned channel is closed.
+func (e *Engine) monitorLease(ctx context.Context, lease influxdb.Lease) chan struct{} {
+	done := make(chan struct{})
+	go func(lease influxdb.Lease) {
+		ttl, err := lease.TTL(ctx)
+		if err != nil {
+			e.logger.Warn("unable to get TTL for lease on semaphore", zap.Error(err))
+			ttl = influxdb.DefaultLeaseTTL // This is probably a reasonable fallback.
+		}
+
+		// Renew the lease when ttl is halved
+		ticker := time.NewTicker(ttl / 2)
+		for {
+			select {
+			case <-done:
+				ticker.Stop()
+				if err := lease.Release(ctx); err != nil {
+					e.logger.Warn("Lease on sempahore was not released", zap.Error(err))
+				}
+				return
+			case <-ticker.C:
+				if err := lease.KeepAlive(ctx); err != nil {
+					e.logger.Warn("Unable to extend lease", zap.Error(err))
+				} else {
+					e.logger.Info("Extended lease on semaphore")
+				}
+			}
+		}
+	}(lease)
+	return done
+}
```

Comment:
I was thinking that `monitorLease()` wouldn't have a goroutine and that the caller would do `go e.monitorLease()`.

Also, it seems like using `context.WithCancel()` and using `ctx` instead of `done` would be more idiomatic.